File: src/test/java/com/raychenon/leetcode/ShortestWaytoFormStringTest.java
package com.raychenon.leetcode;

import org.junit.BeforeClass;
import org.junit.Test;

import static junit.framework.TestCase.assertEquals;

/**
 * User: raychenon
 * Date: 2019-10-14
 * https://leetcode.com/problems/shortest-way-to-form-string/
 */
public class ShortestWaytoFormStringTest {

    private static ShortestWaytoFormString shortestWaytoFormString;

    @BeforeClass
    public static void setUp() {
        shortestWaytoFormString = new ShortestWaytoFormString();
    }

    @Test
    public void example1Test() {
        assertShortestWay(2, "abc", "abcbc");
    }

    @Test
    public void example2Test() {
        assertShortestWay(-1, "abc", "acdbc");
    }

    @Test
    public void example3Test() {
        assertShortestWay(3, "xyz", "xzyxz");
    }

    void assertShortestWay(int expectedValue, String source, String target) {
        assertEquals(expectedValue, shortestWaytoFormString.shortestWayPointers(source, target));
        assertEquals(expectedValue, shortestWaytoFormString.shortestWayInvertedIndex(source, target));
    }

}

File: src/test/kotlin/com/raychenon/kotlin/leetcode/ShortestWaytoFormStringTest.kt
package com.raychenon.kotlin.leetcode

import org.junit.Assert
import org.junit.Test

/**
 * User: raychenon
 * Date: 2019-10-13
 * https://leetcode.com/problems/shortest-way-to-form-string/
 */
class ShortestWaytoFormStringTest {

    @Test
    fun example1Test() {
        assertShortestWayMethods(2, "abc", "abcbc")
    }

    @Test
    fun example2Test() {
        assertShortestWayMethods(-1, "abc", "acdbc")
    }

    @Test
    fun example3Test() {
        assertShortestWayMethods(3, "xyz", "xzyxz")
    }

    @Test
    fun duplicateCharInSourceTest() {
        // aab ab ab aaa a
        assertShortestWayMethods(5, "aaab", "aabababaaaa")
    }
    
    fun assertShortestWayMethods(expected: Int, source: String, target: String) {
        Assert.assertEquals(expected, ShortestWaytoFormString.shortestWayPointers(source, target))
        Assert.assertEquals(expected, ShortestWaytoFormString.shortestWayInvertedIndex(source, target))
    }
}